### PR TITLE
Asavioli perfstat snapshot - make changes to stats resultsets and replace * with explicit columns in queries

### PR DESCRIPTION
**Pull request by asavioli**
Fixes # .

Changes proposed in this pull request:
 - issue #125 
   - merged the Perf Stats snapshot file with some improvements coming from SQL Log Scout
     -  replaced some select * by select with the column list
     - improvement for the part collecting data about the statistics (--new stats DMV). Instead of executing multiple statements to check if a database is an AG secondary replica with readable secondary disabled or read-intent only, the function HAS_PERMS_BY_NAME is being used
   
   - improvements not related to the SQL Log Scout version
     - improvement for the part collecting data about the statistics (--new stats DMV): schema name being returned, not only the object name
     - also included the name of the statistic on the result set which was missing
     - changed the filter to return statistics from master and msdb database (before the filter was database_id >4)
   - issue #140 
     - removed duplicated part collecting statistics information  

 

How to test this code:
 - 
 - 
 - 

Has been tested on (remove any that don't apply):
 - Case-sensitive SQL Server instance
 - SQL Server 2008
 - SQL Server 2008 R2
 - SQL Server 2012
 - SQL Server 2014
 - SQL Server 2016
 - SQL Server 2017
 - SQL Server 2019
 - Azure SQL Virtual Machine
 - Amazon RDS

### CODE REVIEW

**Comment by PiJoCoder** [comment]
Tested this - data gets collected successfully. See comment about making a change

File: DiagManager/CustomDiagnostics/SQL Server Perf Stats/SQL Server Perf Stats Snapshot.sql [hunk 1]
```diff
@@ -164,107 +164,95 @@ begin
...
-		set @sql = @sql + '	insert into #tmpStats	
-		                    select ' + cast( @dbid as nvarchar(20)) +   ' ''Database_Id''' + ',''' +  @dbname  + ''' Database_Name,  Object_name(st.object_id) ''Object_Name'',  st.* 
-		                    from sys.stats ss  
-							  cross apply sys.dm_db_stats_properties (ss.object_id, ss.stats_id) st 
-		                      inner join sys.objects so ON (ss.object_id = so.object_id) where so.type_desc not in (''SYSTEM_TABLE'', ''INTERNAL_TABLE'')'  
-							  --replaced sys.dm_db_index_usage_stats  by sys.stat since the first doesn't return anything in case the table or index was not accessed since last SQL restart
-							  
-		exec (@sql)
+	    --replaced sys.dm_db_index_usage_stats  by sys.stat since the first doesn't return anything in case the table or index was not accessed since last SQL restart
+		set @sql = @sql + '	insert into #tmpStats	select ' + cast( @dbid as nvarchar(20)) +   ' ''Database_Id''' + ',''' +  @dbname  + ''' Database_Name,  Object_name(st.object_id) ''Object_Name'', SCHEMA_NAME(schema_id) ''Schema_Name'', st.* from sys.stats ss
```
Comment (PiJoCoder) [inline DiagManager/CustomDiagnostics/SQL Server Perf Stats/SQL Server Perf Stats Snapshot.sql]: Can you consider adding the columns explicitly and replace * in this query 
  ` st.* from sys.stats ss `
That way you control what SQL Nexus imports. Also check for any other places in this stats query that you replaced.

Comment (asavioli) [reply DiagManager/CustomDiagnostics/SQL Server Perf Stats/SQL Server Perf Stats Snapshot.sql]: Thanks for the heads up! The modification has been performed

**Comment by PiJoCoder** [comment]
@asavioli - this looks great. Could you re-compute the Hash for the Perfstat snapshot sql file and update in the Confirm*.ps1 file. Then commit and this will be ready for merge

image

File: DiagManager/CustomDiagnostics/SQL Server Perf Stats/SQL Server Perf Stats Snapshot.sql [hunk 2]
```diff
@@ -225,64 +225,119 @@ begin
 
 	--new stats DMV
 	set nocount on
-	declare @dbname sysname, @dbid int	
+	declare @dbname sysname, @dbid int,  @sql_major_version INT, @sql_major_build INT
+
+	SELECT @sql_major_version = (LEFT(CAST(SERVERPROPERTY(N'ProductVersion') AS NVARCHAR(128)), (CHARINDEX(N'.', CAST(SERVERPROPERTY(N'ProductVersion') AS NVARCHAR(128)))-1))),
```
Comment (PiJoCoder) [inline DiagManager/CustomDiagnostics/SQL Server Perf Stats/SQL Server Perf Stats Snapshot.sql:230]: Could we replace this line with something like this:

SELECT @sql_major_version = (CAST(PARSENAME(CAST(SERVERPROPERTY('ProductVersion') AS varchar(20)), 4) AS INT)) 

It is going to be consistent with the second line, it casts the value directly into INT as opposed to relying on an implicit cast and is overall more efficient.

**Comment by PiJoCoder** [comment]
Also, let's discuss the other issue that you found with converting the columns explicitly.

Unfortunately this is coming from our change here where we are expanding the "*" into explicit column names. For example that column that in the first error was introduced in SQL 2014

https://docs.microsoft.com/en-us/sql/relational-databases/system-catalog-views/sys-resource-governor-configuration-transact-sql?view=sql-server-ver15#:~:text=max_outstanding_io_per_volume,O%20per%20volume.

So we have a decision to make 
- either we go back from column list to * for these tables and let SQL Nexus figure out any outstanding columns and import them as varchar 
- or we have to do this version checking that you just introduced for these columns as well

Changed database context to 'tempdb'.
Changed database context to 'tempdb'.
Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot10, Line 15
Invalid column name 'max_outstanding_io_per_volume'.
Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot10, Line 28
Invalid column name 'min_iops_per_volume'.
Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot10, Line 29
Invalid column name 'max_iops_per_volume'.
Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot10, Line 44
Invalid column name 'external_pool_id'.
Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot11, Line 56
Invalid column name 'version'.
Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot11, Line 57
Invalid column name 'basic_features'.
Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot11, Line 58
Invalid column name 'dtc_support'.
Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot11, Line 59
Invalid column name 'db_failover'.
Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot11, Line 60
Invalid column name 'is_distributed'.
Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot11, Line 121
Invalid column name 'seeding_mode'.
Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot11, Line 122
Invalid column name 'seeding_mode_desc'.
Msg 2812, Level 16, State 62, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_Run_PerfStats_Snapshot______________________________________________________________________________________________00000718, Line 8
Could not find stored procedure '#sp_perf_stats_snapshot11'.
executing procedure #sp_perf_stats_snapshot11

**Comment by asavioli** [comment]
> Also, let's discuss the other issue that you found with converting the columns explicitly.
> 
> 
> 
> Unfortunately this is coming from our change here where we are expanding the "*" into explicit column names. For example that column that in the first error was introduced in SQL 2014
> 
> 
> 
> https://docs.microsoft.com/en-us/sql/relational-databases/system-catalog-views/sys-resource-governor-configuration-transact-sql?view=sql-server-ver15#:~:text=max_outstanding_io_per_volume,O%20per%20volume.
> 
> 
> 
> So we have a decision to make 
> 
> - either we go back from column list to * for these tables and let SQL Nexus figure out any outstanding columns and import them as varchar 
> 
> - or we have to do this version checking that you just introduced for these columns as well
> 
> 
> 
> Changed database context to 'tempdb'.
> 
> Changed database context to 'tempdb'.
> 
> Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot10, Line 15
> 
> Invalid column name 'max_outstanding_io_per_volume'.
> 
> Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot10, Line 28
> 
> Invalid column name 'min_iops_per_volume'.
> 
> Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot10, Line 29
> 
> Invalid column name 'max_iops_per_volume'.
> 
> Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot10, Line 44
> 
> Invalid column name 'external_pool_id'.
> 
> Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot11, Line 56
> 
> Invalid column name 'version'.
> 
> Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot11, Line 57
> 
> Invalid column name 'basic_features'.
> 
> Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot11, Line 58
> 
> Invalid column name 'dtc_support'.
> 
> Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot11, Line 59
> 
> Invalid column name 'db_failover'.
> 
> Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot11, Line 60
> 
> Invalid column name 'is_distributed'.
> 
> Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot11, Line 121
> 
> Invalid column name 'seeding_mode'.
> 
> Msg 207, Level 16, State 1, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_perf_stats_snapshot11, Line 122
> 
> Invalid column name 'seeding_mode_desc'.
> 
> Msg 2812, Level 16, State 62, Server DESKTOP-2933TUN\SQL2012, Procedure #sp_Run_PerfStats_Snapshot______________________________________________________________________________________________00000718, Line 8
> 
> Could not find stored procedure '#sp_perf_stats_snapshot11'.
> 
> executing procedure #sp_perf_stats_snapshot11

I would go for the second option. Since this is a simple small select, let's do the same version checking. Will change it on Monday along with the other change you suggested on the CAST

**Comment by PiJoCoder** [comment]
> I would go for the second option. Since this is a simple small select, let's do the same version checking. Will change it on Monday along with the other change you suggested on the CAST

Sounds good. Just bear in mind that we are talking about multiple SELECTs here - each error 207 is raised in a different table, or at least most of them.

**Comment by PiJoCoder** [comment]
@asavioli , thanks. One suggestion to consider, it may be better for performance to avoid dynamically-constructed T-SQL, though it this case since there are no parameters, it may not matter. But I would suggest simpler approach that you used for the previous query, though it would take up more lines of code:

 IF (@sql_major_version >=11)
   SELECT ... entire query here, with the extra column
 ELSE
  SELECT - older version without the column
  

Thoughts?

**Comment by asavioli** [comment]
> @asavioli , thanks. One suggestion to consider, it may be better for performance to avoid dynamically-constructed T-SQL, though it this case since there are no parameters, it may not matter. But I would suggest simpler approach that you used for the previous query, though it would take up more lines of code:
> 
> 
> 
>  IF (@sql_major_version >=11)
> 
>    SELECT ... entire query here, with the extra column
> 
>  ELSE
> 
>   SELECT - older version without the column
> 
>   
> 
> 
> 
> Thoughts?
> 
> 
> 
> 
> 
> 

Hey Joseph. 

This was us my first attempt, but I started getting errors about not existing columns, even having an IF. I think SQL checks it during compilation time. That's why I went to the dynamic queries

**Review by PiJoCoder — APPROVED** [review]

**Pull request merged**